### PR TITLE
Add safety checks to `samMethod`

### PR DESCRIPTION
**Pull request by anatoliykmetyuk**
`sam` in `samMethod` stands for "single abstract method". However,
presently, no check is performed to verify that the abstract method
it finds is indeed single. This commit adds such a check to this method.

As mentioned [here](https://github.com/lampepfl/dotty/pull/6266#issuecomment-481722323), `samMethod` is unsafe. Since it is only used in [`genInvokeDynamicLambda`](https://github.com/lampepfl/dotty/blob/e750a8afbba8b2448b5bb4b9728c6129911ca8cb/compiler/src/dotty/tools/backend/jvm/BCodeBodyBuilder.scala#L1439), which is only used when [generating the bytecode](https://github.com/lampepfl/dotty/blob/e750a8afbba8b2448b5bb4b9728c6129911ca8cb/compiler/src/dotty/tools/backend/jvm/BCodeBodyBuilder.scala#L316) for `Closure`s, this is no immediate problem. However, since this method is a part of a public API, not performing the safety checks can lead to runtime errors if it is used elsewhere incorrectly.

### CODE REVIEW

File: compiler/src/dotty/tools/backend/jvm/DottyBackendInterface.scala [hunk 1]
```diff
@@ -853,8 +853,12 @@ class DottyBackendInterface(outputDirectory: AbstractFile, val superCallsMap: Ma
 
     def addRemoteRemoteExceptionAnnotation: Unit = ()
 
-    def samMethod(): Symbol =
-      toDenot(sym).info.abstractTermMembers.headOption.getOrElse(toDenot(sym).info.member(nme.apply)).symbol
+    def samMethod(): Symbol = {
+      val abstractTermMembers = toDenot(sym).info.abstractTermMembers
+      if (abstractTermMembers.size > 1) abort(s"${sym.show} is not a functional interface. " +
```
Comment (smarter) [inline compiler/src/dotty/tools/backend/jvm/DottyBackendInterface.scala]: I would pattern-match on `abstractTermMembers` and have cases for it being empty, one element and multiple elements.

**Comment by smarter** [comment]
There's a failure in the community build:

> java.lang.RuntimeException: trait JProcedure1 is not a functional interface. It has the following abstract methods: apply, applyVoid

This happens because dotty-library and the 2.13 standard library are being jointly compiled, so we don't have any bridge for JProcedure1 since it's a .java source, this illustrates exactly the issue I predicted in https://github.com/lampepfl/dotty/pull/6266#issuecomment-482800332

**Comment by anatoliykmetyuk** [comment]
Why does the community build fail but nothing else; what is special about the community build?

The two deferred methods are a problem though since `JProcedure1` is no longer a functional interface.

**Comment by smarter** [comment]
In the community build, we compile the 2.13 standard library, but because we can't just mix a dotty-library compiled with the 2.12 standard library and the 2.13 standard library, we need to compile the dotty-library at the same time, so we end up jointly compiling the sources of the 2.13 standard library and the sources of dotty-library. Hence .java files go through the JavaParser then their compilation unit gets dropped after typer, so we don't see any bridge for them. The fix to samMethod I suggested in https://github.com/lampepfl/dotty/pull/6266#issuecomment-482802119 should fix this.

**Comment by anatoliykmetyuk** [comment]
Makes sense. Ok, let's try...

Btw, what's the correct procedure to run the community tests? Locally I am getting different errors but not the one as in the CI

**Comment by smarter** [comment]
`community-build/test` should work.

**Comment by anatoliykmetyuk** [comment]
Oh, this time it worked...

Hmm, now I'm getting the errors of the following kind:

```scala
[error] trait JFunction1$mcII$sp is not a functional interface. It has the following abstract methods: apply, apply$mcII$spjava.lang.RuntimeException: trait JFunction1$mcII$sp is not a functional interface. It has the following abstract methods: apply, apply$mcII$sp

	at dotty.tools.backend.jvm.DottyBackendInterface.abort(DottyBackendInterface.scala:390)
	at dotty.tools.backend.jvm.DottyBackendInterface$$anon$8.$anonfun$samMethod$1(DottyBackendInterface.scala:861)
	at dotty.tools.dotc.core.Periods.atPhase(Periods.scala:25)
	at dotty.tools.dotc.core.Phases.atPhase(Phases.scala:35)
	at dotty.tools.dotc.core.Phases.atPhase$(Phases.scala:34)
	at dotty.tools.dotc.core.Contexts$Context.atPhase(Contexts.scala:71)
	at dotty.tools.backend.jvm.DottyBackendInterface$$anon$8.samMethod(DottyBackendInterface.scala:856)
	at dotty.tools.backend.jvm.DottyBackendInterface$$anon$8.samMethod(DottyBackendInterface.scala:632)
	at dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genInvokeDynamicLambda(BCodeBodyBuilder.scala:1472)
	at dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genLoad(BCodeBodyBuilder.scala:335)
	at dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genLoad(BCodeBodyBuilder.scala:408)
	at dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.$anonfun$genLoadArguments$1(BCodeBodyBuilder.scala:1051)
	at dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.$anonfun$genLoadArguments$1$adapted(BCodeBodyBuilder.scala:1051)
	at scala.collection.immutable.List.foreach(List.scala:392)
```

**Comment by anatoliykmetyuk** [comment]
It seems that we're indeed relying on the bridges in the joint compilation after all. And we're indeed having a bug there which was not previously exposed due to the `samMethod` not performing safety checks and by coincidence picking up the right method. Each time, for some reason...

**Comment by smarter** [comment]
> trait JFunction1$mcII$sp is not a functional interface. It has the following abstract methods: apply, apply$mcII$sp

Just pushed a commit that fixed this.

**Comment by anatoliykmetyuk** [comment]
So no bridges are needed since the types are the same after erasure? Neat!

**Comment by smarter** [comment]
There's one remaining test failure but it's very weird and I don't have time to investigate right now, @nicolasstucki it involves quoting so I'm wondering if have you seen something like this before ?
```scala
Test tests/run-with-compiler/i3876-c.scala failed with output:                
6
Exception in thread "main" java.lang.reflect.InvocationTargetException
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at dotty.tools.vulpix.ChildJVMMain.runMain(ChildJVMMain.java:40)
        at dotty.tools.vulpix.ChildJVMMain.main(ChildJVMMain.java:47)
Caused by: java.lang.ArrayIndexOutOfBoundsException: 36
        at dotty.tools.dotc.core.Periods.hasSameBaseTypesAs(Periods.scala:47)
        at dotty.tools.dotc.core.SymDenotations$ClassDenotation.baseTypeCache(SymDenotations.scala:1408)
        at dotty.tools.dotc.core.SymDenotations$ClassDenotation.baseTypeOf(SymDenotations.scala:1730)
        at dotty.tools.dotc.core.Types$Type.baseType(Types.scala:925)
        at dotty.tools.dotc.core.TypeOps$AsSeenFromMap.toPrefix$1(TypeOps.scala:46)
        at dotty.tools.dotc.core.TypeOps$AsSeenFromMap.apply(TypeOps.scala:65)
        at dotty.tools.dotc.core.TypeOps$AsSeenFromMap.apply(TypeOps.scala:63)
        at dotty.tools.dotc.core.TypeOps$AsSeenFromMap.apply(TypeOps.scala:35)
        at scala.collection.immutable.List.mapConserve(List.scala:179)
        at dotty.tools.dotc.core.Types$TypeMap.mapOverLambda$1(Types.scala:4437)
        at dotty.tools.dotc.core.Types$TypeMap.mapOver(Types.scala:4441)
        at dotty.tools.dotc.core.TypeOps$AsSeenFromMap.apply(TypeOps.scala:69)
        at dotty.tools.dotc.core.TypeOps.asSeenFrom(TypeOps.scala:30)
        at dotty.tools.dotc.core.Contexts$Context.asSeenFrom(Contexts.scala:71)
        at dotty.tools.dotc.core.Types$Type.asSeenFrom(Types.scala:832)
        at dotty.tools.dotc.core.Denotations$SingleDenotation.computeAsSeenFrom(Denotations.scala:1118)
        at dotty.tools.dotc.core.Denotations$SingleDenotation.computeAsSeenFrom(Denotations.scala:1111)
        at dotty.tools.dotc.core.Denotations$PreDenotation.asSeenFrom(Denotations.scala:128)
        at dotty.tools.dotc.core.SymDenotations$ClassDenotation.findMember(SymDenotations.scala:1725)
        at dotty.tools.dotc.core.Types$Type.go$1(Types.scala:562)
        at dotty.tools.dotc.core.Types$Type.goRefined$1(Types.scala:651)
        at dotty.tools.dotc.core.Types$Type.go$1(Types.scala:581)
        at dotty.tools.dotc.core.Types$Type.findMember(Types.scala:712)
        at dotty.tools.dotc.core.Types$Type.memberBasedOnFlags(Types.scala:545)
        at dotty.tools.dotc.core.Types$Type.member(Types.scala:530)
        at dotty.tools.dotc.typer.TypeAssigner.selectionType(TypeAssigner.scala:238)
        at dotty.tools.dotc.typer.Typer.selectionType(Typer.scala:87)
        at dotty.tools.dotc.typer.TypeAssigner.accessibleSelectionType(TypeAssigner.scala:268)
        at dotty.tools.dotc.typer.Typer.accessibleSelectionType(Typer.scala:87)
        at dotty.tools.dotc.typer.TypeAssigner.assignType(TypeAssigner.scala:299)
        at dotty.tools.dotc.typer.Typer.assignType(Typer.scala:87)
        at dotty.tools.dotc.ast.tpd$.Select(tpd.scala:30)
        at dotty.tools.dotc.ast.tpd$TreeOps$.select$extension3(tpd.scala:794)
        at dotty.tools.dotc.core.quoted.PickledQuotes$.rec$1(PickledQuotes.scala:148)
        at dotty.tools.dotc.core.quoted.PickledQuotes$.rec$1(PickledQuotes.scala:146)
        at dotty.tools.dotc.core.quoted.PickledQuotes$.rec$1(PickledQuotes.scala:136)
        at dotty.tools.dotc.core.quoted.PickledQuotes$.functionAppliedTo(PickledQuotes.scala:150)
        at dotty.tools.dotc.core.quoted.PickledQuotes$.quotedExprToTree(PickledQuotes.scala:51)
        at dotty.tools.dotc.quoted.QuoteCompiler$QuotedFrontend.runOn$$anonfun$1(QuoteCompiler.scala:53)
        at scala.collection.immutable.List.map(List.scala:286)
        at dotty.tools.dotc.quoted.QuoteCompiler$QuotedFrontend.runOn(QuoteCompiler.scala:61)
        at dotty.tools.dotc.Run.runPhases$4$$anonfun$4(Run.scala:158)
        at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
        at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
        at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:36)
        at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:33)
        at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:198)
        at dotty.tools.dotc.Run.runPhases$5(Run.scala:170)
        at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:178)
        at dotty.runtime.function.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
        at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:102)
        at dotty.tools.dotc.Run.compileUnits(Run.scala:185)
        at dotty.tools.dotc.Run.compileUnits(Run.scala:125)
        at dotty.tools.dotc.quoted.QuoteCompiler$ExprRun.compileExpr(QuoteCompiler.scala:90)
        at dotty.tools.dotc.quoted.QuoteDriver.withTree(QuoteDriver.scala:73)
        at dotty.tools.dotc.quoted.QuoteDriver.show(QuoteDriver.scala:59)
        at dotty.tools.dotc.quoted.ToolboxImpl$$anon$1.show(ToolboxImpl.scala:31)
        at scala.quoted.Expr.show(Expr.scala:14)
        at Test$.main(i3876-c.scala:13)
        at Test.main(i3876-c.scala)
        ... 6 more

**Comment by nicolasstucki** [comment]
It looks like another instance of #5161

**Pull request merged**